File: dynamodb/src/main/scala/models/Query.scala
package aws.dynamodb

sealed trait KeyCondition

object KeyCondition {

  case class EqualTo(value: DDBAttribute) extends KeyCondition

  case class LessThan(value: DDBAttribute, orEqual: Boolean) extends KeyCondition

  case class GreaterThan(value: DDBAttribute, orEqual: Boolean) extends KeyCondition

  case class BeginsWith(value: DDBAttribute) extends KeyCondition

  case class Between(lowerBound: DDBAttribute, upperBound: DDBAttribute) extends KeyCondition

}

/**
 * A DynamoDB Query.
 *
 * @param tableName The name of the table containing the requested items.
 * @param hashKeyValue Attribute value of the hash component of the composite primary key.
 * @param attributesToGet Array of Attribute names. If Nil then all attributes will be returned.
 *                        If some attributes are not found, they will not appear in the result.
 * @param limit The maximum number of items to return (not necessarily the number of matching items).
 *              If Amazon DynamoDB processes the number of items up to the limit while querying the table,
 *              it stops the query and returns the matching values up to that point,
 *              and a `LastEvaluatedKey` to apply in a subsequent operation to continue the query.
 *              Also, if the result set size exceeds 1MB before Amazon DynamoDB hits this limit,
 *              it stops the query and returns the matching values, and a `LastEvaluatedKey` to apply
 *              in a subsequent operation to continue the query.
 * @param consistentRead If set to `true`, then a consistent read is issued, otherwise eventually consistent is used.
 * @param count If set to true, Amazon DynamoDB returns a total number of items that match the query parameters,
 *              instead of a list of the matching items and their attributes. You can apply the Limit parameter to
 *              count-only queries.
 *              Do not set Count to true while providing a list of AttributesToGet;
 *              otherwise, Amazon DynamoDB returns a validation error.
 *              For more information, see [[http://docs.amazonwebservices.com/amazondynamodb/latest/developerguide/QueryAndScan.html#Count Count and ScannedCount]].
 * @param rangeKeyCondition The attribute values and comparison operators to use for the query.
 *                          If `None`, Amazon DynamoDB returns all items with the specified hash key element value.
 * @param scanIndexForward Specifies ascending or descending traversal of the index.
 *                         Amazon DynamoDB returns results reflecting the requested order determined by the range key,
 *                         based on ASCII character code values. Defaults to `true` (ascending).
 * @param exclusiveStartKey Primary key of the item from which to continue an earlier query. An earlier query might provide
 *                          this value as the LastEvaluatedKey if that query operation was interrupted before completing the query;
 *                          either because of the result set size or the Limit parameter.
 *                          The `LastEvaluatedKey` can be passed back in a new query request to continue the operation
 *                          from that point.
 *
 */
case class Query(tableName: String,
                 hashKeyValue: DDBAttribute,
                 attributesToGet: Seq[String] = Nil,
                 limit: Option[Long] = None,
                 consistentRead: Boolean = false,
                 count: Boolean = false,
                 rangeKeyCondition: Option[KeyCondition] = None,
                 scanIndexForward: Boolean = true,
                 exclusiveStartKey: Option[PrimaryKey] = None)

File: dynamodb/src/main/java/com/pellucid/aws/dynamodb/models/Query.java
package com.pellucid.aws.dynamodb.models;

import java.util.ArrayList;
import java.util.List;

import scala.collection.JavaConversions;

import play.libs.Scala;

public class Query {

    private String tableName;

    private AttributeValue hashKeyValue;

    private List<String> attributesToGet;

    private Long limit;

    private boolean consistentRead;

    private boolean count;

    private boolean scanIndexForward;

    private KeyCondition rangeKeyCondition;

    private PrimaryKey exclusiveStartKey;

    public Query(String tableName) {
        this(tableName, null, new ArrayList<String>(), null, false, false, true, null, null);
    }

    private Query(
            String tableName,
            AttributeValue hashKeyValue,
            List<String> attributesToGet,
            Long limit,
            Boolean consistentRead,
            Boolean count,
            Boolean scanIndexForward,
            KeyCondition rangeKeyCondition,
            PrimaryKey exclusiveStartKey
            ) {
        this.tableName = tableName;
        this.hashKeyValue = hashKeyValue;
        this.attributesToGet = new ArrayList<String>(attributesToGet);
        this.limit = limit;
        this.consistentRead = consistentRead;
        this.count = count;
        this.rangeKeyCondition = rangeKeyCondition;
        this.exclusiveStartKey = exclusiveStartKey;
    }

    public Query withHashKeyValue(AttributeValue hashKeyValue) {
        return new Query(
                this.tableName,
                hashKeyValue,
                this.attributesToGet,
                this.limit,
                this.consistentRead,
                this.count,
                this.scanIndexForward,
                this.rangeKeyCondition,
                this.exclusiveStartKey
                );
    }

    public Query withAttributesToGet(List<String> attributesToGet) {
        return new Query(
                this.tableName,
                this.hashKeyValue,
                attributesToGet,
                this.limit,
                this.consistentRead,
                this.count,
                this.scanIndexForward,
                this.rangeKeyCondition,
                this.exclusiveStartKey
                );
    }

    public Query withLimit(Long limit) {
        return new Query(
                this.tableName,
                this.hashKeyValue,
                this.attributesToGet,
                limit,
                this.consistentRead,
                this.count,
                this.scanIndexForward,
                this.rangeKeyCondition,
                this.exclusiveStartKey
                );
    }

    public Query withConsistentRead(Boolean consistentRead) {
        return new Query(
                this.tableName,
                this.hashKeyValue,
                this.attributesToGet,
                this.limit,
                consistentRead,
                this.count,
                this.scanIndexForward,
                this.rangeKeyCondition,
                this.exclusiveStartKey
                );
    }

    public Query withCount(Boolean count) {
        return new Query(
                this.tableName,
                this.hashKeyValue,
                this.attributesToGet,
                this.limit,
                this.consistentRead,
                count,
                this.scanIndexForward,
                this.rangeKeyCondition,
                this.exclusiveStartKey
                );
    }

    public Query withRangeKeyCondition(KeyCondition rangeKeyCondition) {
        return new Query(
                this.tableName,
                this.hashKeyValue,
                this.attributesToGet,
                this.limit,
                this.consistentRead,
                this.count,
                this.scanIndexForward,
                rangeKeyCondition,
                this.exclusiveStartKey
                );
    }

    public Query withExclusiveStartKey(PrimaryKey exclusiveStartKey) {
        return new Query(
                this.tableName,
                this.hashKeyValue,
                this.attributesToGet,
                this.limit,
                this.consistentRead,
                this.count,
                this.scanIndexForward,
                this.rangeKeyCondition,
                exclusiveStartKey
                );
    }

    public aws.dynamodb.Query toScala() {
        return new aws.dynamodb.Query(
                this.tableName,
                this.hashKeyValue.toScala(),
                JavaConversions.iterableAsScalaIterable(this.attributesToGet).toSeq(),
                Scala.Option((Object)this.limit),
                this.consistentRead,
                this.count,
                Scala.Option(this.rangeKeyCondition == null ? null : this.rangeKeyCondition.toScala()),
                this.scanIndexForward,
                Scala.Option(this.exclusiveStartKey == null ? null : this.exclusiveStartKey.toScala())
                );
    }

}
